publisher/publisher.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2015 ISRG.  All rights reserved
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package publisher
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type LogDescription struct {
+	ID        string
+	URI       string
+	PublicKey *ecdsa.PublicKey
+}
+
+type rawLogDescription struct {
+	URI       string `json:"uri"`
+	PublicKey string `json:"key"`
+}
+
+func (logDesc *LogDescription) UnmarshalJSON(data []byte) error {
+	var rawLogDesc rawLogDescription
+	if err := json.Unmarshal(data, &rawLogDesc); err != nil {
+		return fmt.Errorf("Failed to unmarshal log description, %s", err)
+	}
+	logDesc.URI = rawLogDesc.URI
+	// Load Key
+	pkBytes, err := base64.StdEncoding.DecodeString(rawLogDesc.PublicKey)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	pk, err := x509.ParsePKIXPublicKey(pkBytes)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	var ok bool
+	if logDesc.PublicKey, ok = pk.(*ecdsa.PublicKey); !ok {
+		return fmt.Errorf("Failed to unmarshal log description for %s, unsupported public key type", logDesc.URI)
+	}
+
+	// Generate key hash for log ID
+	pkHash := sha256.Sum256(pkBytes)
+	logDesc.ID = base64.StdEncoding.EncodeToString(pkHash[:])
+	if len(logDesc.ID) != 44 {
+		return fmt.Errorf("Invalid log ID length [%d]", len(logDesc.ID))
+	}
+
+	return nil
+}
+
+// CTConfig defines the JSON configuration file schema
+type CTConfig struct {
+	Logs              []LogDescription `json:"logs"`
+	SubmissionRetries int              `json:"submissionRetries"`
+	// This should use the same method as the DNS resolver
+	SubmissionBackoffString string        `json:"submissionBackoff"`
```

Comment:
I think Go can parse a time.Duration directly out of JSON without needing this intermediate, right?